### DISCUSSION TITLE
FastAPI / SQLAlchemy based book server

### DISCUSSION THREAD

**Discussion by bnmnetp** [RunestoneInteractive/BookServer#13] [General]
This is a start on a migration of the book server to FastAPI - Lets use this Issue to discuss.

I want to polish up this first bit to set the stage for the rest of what needs to come.  So getting this as "right" as we can is the goal.

You should be able to run this yourself with the following:

1. clone this repo
2. Install poetry
3. run `poetry install` to install all of the dependencies (you can either make your own virtualenv or let Poetry)
4. run the command `uvicorn app.main:app --reload`
5. Navigate to http://localhost:8000/docs

You can then exercise the API call for logger/bookevent using the autogenerated documentation.

Next up getting a unittest set up for this first endpoint.

Tagging @bjones1

**Comment by bnmnetp** [comment]
@bjones1 

I think this would be a great time for us to take a look at the bookserver together and see how we feel about organization, complexity, authentication, etc.  

As of today, if you have an entry for a user in auth_user in a database -- I continue to test with sqllite -- and an entry for a course, AND you have built a book with the bookserver branch of the components.  you can login, view the book and page views are all added to the useinfo table.

I am to the point where I think I can really start chopping wood on porting the ajax.py endpoints and getting test coverage up.

Authentication and my busyness with other tasks big roadblock for me, but I think that I'm through that roadblock and we may be within a couple of weeks of having the server at a really great point.

Happy to schedule some time to zoom and walk through things.

**Comment by bjones1** [comment]
Fantastic! Let me play with it a bit then schedule a time to chat. I'm very excited!

**Comment by bjones1** [comment]
See my thoughts in #8.

**Comment by bjones1** [comment]
What's the plan for instructor use of the bookserver?

1. I can imagine a case where an instructor installs just the bookserver (since that's easy), then runs it for himself/herself or for a small class. BUT there's no admin interface / way to assign problems / etc. Our code doesn't run in this case, since it wants to look for a web2py private key. Should we support this mode by automatically creating and saving a private key? We already have a JWT private key -- would it make sense to use the web2py private key instead if it's available, then fall back to this key if not for both JWT and password hashing? Or should these be separate keys for security reasons?
2. The more typical case would be an instructor installs Docker with both web2py and the bookserver. This is the case we're implicitly building.
